payment/controllers.go

```diff
@@ -0,0 +1,353 @@
+				"Error validating request url parameter",
+				map[string]interface{}{
+					"orderID": "orderID must be a uuidv4",
+				},
+			)
+		}
+
+		id := uuid.Must(uuid.FromString(orderID))
+
+		order, err := service.datastore.GetOrder(id)
+		if err != nil {
+			return handlers.WrapError(err, "Error retrieving the order", http.StatusInternalServerError)
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		if order == nil {
+			w.WriteHeader(http.StatusNotFound)
+		} else {
+			w.WriteHeader(http.StatusOK)
+		}
+		if err := json.NewEncoder(w).Encode(order); err != nil {
+			return handlers.WrapError(err, "Error encoding the orders JSON", http.StatusInternalServerError)
+		}
+		return nil
+	})
+}
+
+// GetTransactions is the handler for listing the transactions for an order
+func GetTransactions(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		orderID := chi.URLParam(r, "orderID")
+		if orderID == "" || !govalidator.IsUUIDv4(orderID) {
+			return handlers.ValidationError(
+				"Error validating request url parameter",
+				map[string]interface{}{
+					"orderID": "orderID must be a uuidv4",
+				},
+			)
+		}
+
+		id := uuid.Must(uuid.FromString(orderID))
+
+		order, err := service.datastore.GetTransactions(id)
+		if err != nil {
+			return handlers.WrapError(err, "Error retrieving the transactions for the order", http.StatusInternalServerError)
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		if err := json.NewEncoder(w).Encode(order); err != nil {
+			return handlers.WrapError(err, "Error encoding the transactions JSON", http.StatusInternalServerError)
+		}
+		return nil
+	})
+}
+
+// CreateTransactionRequest includes information needed to create a transaction
+type CreateTransactionRequest struct {
+	ExternalTransactionID string `json:"externalTransactionID" valid:"uuidv4"`
+}
+
+// CreateUpholdTransaction creates a transaction against an order
+func CreateUpholdTransaction(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		var req CreateTransactionRequest
+		err := requestutils.ReadJSON(r.Body, &req)
+		if err != nil {
+			return handlers.WrapError(err, "Error in request body", http.StatusBadRequest)
+		}
+
+		orderID := chi.URLParam(r, "orderID")
+		if orderID == "" || !govalidator.IsUUIDv4(orderID) {
+			return handlers.ValidationError(
+				"Error validating request url parameter",
+				map[string]interface{}{
+					"orderID": "orderID must be a uuidv4",
+				},
+			)
+		}
+		validOrderID := uuid.Must(uuid.FromString(orderID))
+
+		_, err = govalidator.ValidateStruct(req)
+		if err != nil {
+			return handlers.WrapValidationError(err)
+		}
+
+		// Ensure the external transaction ID hasn't already been added to any orders.
+		transaction, err := service.datastore.GetTransaction(req.ExternalTransactionID)
```

Comment:
I'm concerned about the idempotency of this API - if the first request succeeds on the backend but the response fails to reach the client then the second retry request will fail and the client may get stuck in a retry loop

Reply:
Agreed, instead of an `HTTP 500` error I modified to return an `HTTP 409 - Conflict`. The client should be able to check to see if the status is `200`, `409` or `500` during reprocessing.

https://github.com/brave-intl/bat-go/pull/226/commits/ddaa1925b3cd950bdc299c515def5d66dfc5bc41